libraries/chain/include/eosio/chain/host_context.hpp

```diff
@@ -599,6 +601,9 @@ class host_context {
    // act pointer may be invalidated on call to trx_context.schedule_action
    iterator_cache<key_value_object>    keyval_cache;
    std::string                         _pending_console_output;
+
+   const bool                          _is_action    = true;  // used to avoid using dynamic_cast to identify type of host_context
+   const bool                          _is_sync_call = false; // used to avoid using dynamic_cast to identify type of host_context
```

Comment:
An alternative to these would be for `is_action()` and `is_sync_call()` to be virtual and return false on host_context, and then `apply_context` and `sync_call_context` can override the appropriate one to return true. I'm indifferent on which approach to take, just mentioning it.

Reply:
I like your idea. It removes 2 member fields and makes the constructor a little simpler.

https://github.com/AntelopeIO/spring/pull/1296/commits/0b3559cc857cfd90dc735c4b76be489b81410a92